<h1 align="center">An Introduction to Machine Learning - 25737</h1>
<h4 align="center">Dr. Sajjad Amini</h4>
<h4 align="center">Sharif University of Technology, Spring 2023</h4>

**Student Name**:  Amirreza Velaei

**Student ID**:400102222

# Classification

In this question, we will examine classification algorithms. We do not need to implement the algorithms; instead, we will use the functions available in the `scikit-learn` library to solve the question.

## Importing Libraries

First, we import all required libraries.

**Attention**: You should only use these libraries. Other libraries are not acceptable.

In [2]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB

## Loading Data

In this question, we will use the **Breast Cancer** dataset from `scikit-learn`. You can run the following cell to load the dataset:

In [3]:
data = load_breast_cancer() # loading cancer data

X = data.data # features
Y = data.target # labels 

# printing size of the features and targets
print(X.shape)
print(Y.shape)

(569, 30)
(569,)


Now you should use `train_test_split` function to split dataset into three parts:

- 70% for the training set
- 20% for the validation set
- 10% for the test set

Do this in the cell below.

In [4]:
X_train, Y_train, X_val, Y_val, X_test, Y_test = None, None, None, None, None, None

### START CODE HERE ###
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=0)
X_train, X_val, Y_train, Y_val = train_test_split(X_train, Y_train, test_size=0.222, random_state=0)
### END CODE HERE ###

## Logistic Regression

In this part of the question, you will use the `LogisticRegression` classifier to classify the data. You should change the value of `C`, which is the inverse of the regularization parameter, and find the accuracy on the validation set. Using the validation accuracy, you can find the best value for `C`. Note that you may want to change the solver for faster convergence. The `liblinear` solver is recommended for this problem. You can visit [this link](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) for more information about the `LogisticRegression` classifier.

**Question**: What is the best value of `C`?

**Answer**:

In [25]:
### START CODE HERE ###
# Logistic Regression with different values of C
train_acc = []
val_acc = []
for c in [0.01, 0.1, 1, 2, 3, 4, 5,10, 20, 30, 40, 50 , 60 , 70 , 80 , 90 , 100, 1000, 10000, 100000, 1000000]:
    lr = LogisticRegression(C=c)
    lr.fit(X_train, Y_train)
    print("C = ", c, "Training accuracy = ", lr.score(X_train, Y_train), "Validation accuracy = ", lr.score(X_val, Y_val))
    train_acc.append(lr.score(X_train, Y_train))
    val_acc.append(lr.score(X_val, Y_val))

#find the best value of C as min of val_acc
best_c = None
best_val_acc = None
best_train_acc = None

best_c = 0.01
best_val_acc = 0.0
best_train_acc = 0.0
for i in range(len(train_acc)):
    if val_acc[i] > best_val_acc:
        best_val_acc = val_acc[i]
        best_train_acc = train_acc[i]
        best_c = [0.01, 0.1, 1, 2, 3, 4, 5,10, 20, 30, 40, 50 , 60 , 70 , 80 , 90 , 100, 1000, 10000, 100000, 1000000][i]

print("Best C = ", best_c, "Best training accuracy = ", best_train_acc, "Best validation accuracy = ", best_val_acc)
### END CODE HERE ###

d:\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
d:\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_resu

C =  0.01 Training accuracy =  0.9296482412060302 Validation accuracy =  0.9298245614035088
C =  0.1 Training accuracy =  0.9447236180904522 Validation accuracy =  0.956140350877193
C =  1 Training accuracy =  0.9396984924623115 Validation accuracy =  0.9736842105263158
C =  2 Training accuracy =  0.9547738693467337 Validation accuracy =  0.9736842105263158
C =  3 Training accuracy =  0.949748743718593 Validation accuracy =  0.9736842105263158
C =  4 Training accuracy =  0.9371859296482412 Validation accuracy =  0.9736842105263158
C =  5 Training accuracy =  0.949748743718593 Validation accuracy =  0.9736842105263158
C =  10 Training accuracy =  0.9522613065326633 Validation accuracy =  0.9736842105263158
C =  20 Training accuracy =  0.9547738693467337 Validation accuracy =  0.9736842105263158
C =  30 Training accuracy =  0.9522613065326633 Validation accuracy =  0.9736842105263158
C =  40 Training accuracy =  0.9522613065326633 Validation accuracy =  0.9736842105263158
C =  50 Trainin

d:\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
d:\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_resu

Now use your best classifier to calculate the accuracy on the test set.

**Qusetion**: What is the accuracy of this classifier on test set?

**Answer**: Best `C` =  $70$ Best `training accuracy` =  $0.9447236180904522$ Best `validation accuracy` =  $0.9824561403508771$

In [26]:
### START CODE HERE ###
# Logistic Regression with best value of C
lr = LogisticRegression(C=best_c)
lr.fit(X_train, Y_train)
print("Test accuracy = ", lr.score(X_test, Y_test))
### END CODE HERE ###

Test accuracy =  0.9298245614035088


d:\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


For this additional part, you can use only the first two features of the dataset to classify the data and then draw the data points and the boundary on a figure. You can use the [tutorial](https://aleksandarhaber.com/solve-classification-problems-in-python-scikit-learn-and-visualize-the-classification-results-machine-learning-tutorial/?utm_source=rss&utm_medium=rss&utm_campaign=solve-classification-problems-in-python-scikit-learn-and-visualize-the-classification-results-machine-learning-tutorial) provided to learn how to do this. You may need to use additional libraries such as `matplotlib` specifically for this part.

In [7]:
### START CODE HERE ###
# use only the first two features of the dataset to classify the data and then draw the data points and the boundary on a figure
X_train, X_test, Y_train, Y_test = train_test_split(X[:, :2], Y, test_size=0.1, random_state=0)
lr = LogisticRegression(C=best_c)
lr.fit(X_train, Y_train)
print("Test accuracy = ", lr.score(X_test, Y_test))

import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

h = .02  # step size in the mesh

# Create color maps
cmap_light = ListedColormap(['#FFAAAA', '#AAFFAA'])
cmap_bold = ListedColormap(['#FF0000', '#00FF00'])

x_min, x_max = X[:, 0].min() - .5, X[:, 0].max() + .5
y_min, y_max = X[:, 1].min() - .5, X[:, 1].max() + .5

xx, yy = np.meshgrid(np.arange(x_min, x_max, h), np.arange(y_min, y_max, h))

Z = lr.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)

plt.figure()
plt.pcolormesh(xx, yy, Z, cmap=cmap_light)



### END CODE HERE ###


## Naive Bayes

In this part, you will use the `GaussianNB` classifier to classify the data. You should not change the default parameters of this classifier. First, train the classifier on the training set and then find the accuracy of it on the test set.

**Question**: What is the accuracy of this method on test set?

**Answer**:

In [8]:
### START CODE HERE ###

### END CODE HERE ###

## LDA (Linear Discriminant Analysis)

In this part, you will use the `LinearDiscriminantAnalysis` classifier to classify the data. You should not change the default parameters of this classifier. First, train the classifier on the training set and then find the accuracy of it on the test set.

**Question**: What is the accuracy of this method on test set?

**Answer**:

In [9]:
### START CODE HERE ###

### END CODE HERE ###

## Conclution

**Question**: What is the best method for classifying this dataset? What is the best accuracy on the test set?

**Answer**: